src/Compression/CompressionCodecEncrypted.cpp

```diff
@@ -11,204 +19,371 @@
-                                    ErrorCodes::ILLEGAL_SYNTAX_FOR_CODEC_TYPE);
-
-                const String cipher = literal->value.safeGet<String>();
-                if (cipher == "AES-128-GCM-SIV")
-                    return std::make_shared<CompressionCodecEncrypted>(cipher);
-                else
-                    throw Exception("Cipher '" + cipher + "' is not supported",
-                                    ErrorCodes::ILLEGAL_CODEC_PARAMETER);
-            }
-            else
-            {
-                /* The factory is asking us to construct the codec
-                 * only from the method code. How can that be
-                 * possible? For now we only support a single cipher
-                 * so it's not really a problem, but if we were to
-                 * support more ciphers it would be catastrophic. */
-                return std::make_shared<CompressionCodecEncrypted>("AES-128-GCM-SIV");
-            }
-        });
+        key = current_params->keys_storage[method].at(key_id);
     }
+    catch (...)
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "There is no key {} in config", key_id);
+    }
+    nonce = current_params->nonce[method];
+    if (nonce.empty())
+        nonce = {"\0\0\0\0\0\0\0\0\0\0\0\0", 12};
+}
+
+
+CompressionCodecEncrypted::CompressionCodecEncrypted(EncryptionMethod Method): encryption_method(Method)
+{
+    setCodecDescription(getMethodName(encryption_method));
+}
+
+uint8_t CompressionCodecEncrypted::getMethodByte() const
+{
+    return getMethodCode(encryption_method);
+}
+
+void CompressionCodecEncrypted::updateHash(SipHash & hash) const
+{
+    getCodecDesc()->updateTreeHash(hash);
+}
+
+UInt32 CompressionCodecEncrypted::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    // The GCM mode is a stream cipher. No paddings are
+    // involved. There will be a tag at the end of ciphertext (16
+    // octets). Also it has not more than 8 bytes for key_id in the beginning
+    return uncompressed_size + tag_size + key_id_max_size;
+}
+
+UInt32 CompressionCodecEncrypted::doCompressData(const char * source, UInt32 source_size, char * dest) const
+{
+    // Generate an IV out of the data block and the key-generation
+    // key. It is completely deterministic, but does not leak any
+    // information about the data block except for equivalence of
+    // identical blocks (under the same master key). The IV will
+    // be used as an authentication tag. The ciphertext and the
+    // tag will be written directly in the dest buffer.
+    const std::string_view plaintext = std::string_view(source, source_size);
+
+    UInt64 current_key_id;
+    String current_key, nonce;
+    Configuration::instance().getCurrentKeyAndNonce(encryption_method, current_key_id, current_key, nonce);
+
+    char* ciphertext = writeVarUInt(current_key_id, dest);
+    size_t keyid_size = ciphertext - dest;
```

Comment:
Please don't forget to write 1 byte plus maybe a nonce here, as we have decided.